src/cfulist.c

```diff
@@ -412,6 +412,33 @@ cfulist_pop(cfulist_t *list) {
 	return NULL;
 }
 
+void
+cfulist_delete_data(cfulist_t *list, void *data) {
+	cfulist_entry *ptr = NULL;
+
+	if (!list) {
+		return;
+	}
+
+	lock_list(list);
+
+	if (list->entries) {
+		for (ptr = list->entries; ptr && ptr->data != data; ptr = ptr->next)
+		;
+		
+		if (ptr && ptr->data == data) {
+			if (!ptr->prev) {
+				cfulist_dequeue (list);
```

Comment:
Not at all sure about using this function, but I guess it works.  I think you should probably write cfulist_delete_listhead which just calls cfulist_shift or whatever it does, exactly as dequeue does.  Perhaps I'm just being picky, but I feel you're using a queue method for a linked list that just so happens to work for now.

Reply:
I used that function because the doc suggests that its purpose is to remove the first item of a cfulist: ["Remove the value at the beginning of the list (*equivalent to shift*)"](http://libcfu.sourceforge.net/libcfu.html#SEC3)

Also, that function ultimately calls this https://github.com/diogogithub/libcfu/blob/5af4f108dc492a1df99b9c6268c830d3fba62264/src/cfulist.c#L206

Reply:
Do you think it would be better to duplicate this function for better semantics?

Reply:
An argument for not calling another function in this API is that this list seems to have mutex locking/unlocking at very small granularity and it doesn't appear to use a recursive mutex.

A few more things to check:

* The conditional on L425 appears unnecessary as L426 and L429 already guard against `NULL` `entries`.
* The `entries` and `tail` pointers in the list don't appear to be updated for when the `ptr` is first/last.
* The `free_fn` (if set) isn't called on the `data` to clean it up.

Reply:
I'm not familiar with the API here (this repo is a fork and I haven't looked at it in years), but it appears like this list implementation allows the caller to pass [a function that will be called to cleanup the `data`](https://github.com/codebrainz/libcfu/blob/master/doc/libcfu.info#L176) when it is being removed (explicitly, replaced or when the list is destroyed). You can imagine if you were storing `char*` strings in the list in heap memory, you might pass the stdlib `free()` function as the `free_fn` or if you were storing some kind of reference counted items, you might pass the "unref" function used to decrement the item's reference count. Basically it lets the list "automatically" manage the memory of the items stored in it.

Reply:
Ah, `free_fn` from `free_function`, yes that does make sense! I can handle that :)

Reply:
It seems like a cfulist_set_free_function should be created...

Reply:
Also, I think this function's documentation seems to be wrong: https://github.com/diogogithub/libcfu/blob/4d6a4c0a1603e52a5e89119f4134a95b08550ca5/doc/libcfu.texi#L592
Shouldn't it be named [cfulist_destroy_with_free_fn](https://github.com/diogogithub/libcfu/blob/4d6a4c0a1603e52a5e89119f4134a95b08550ca5/src/cfulist.c#L584) ?